crates/store/re_log_encoding/src/protobuf/decoder.rs

```diff
@@ -0,0 +1,136 @@
+    pub(crate) fn decode(data: &mut impl std::io::Read) -> Result<Self, DecodeError> {
+        let kind = MessageKind::decode(data)?;
+        let mut buf = [0; 4];
+        data.read_exact(&mut buf)?;
+        let len = u32::from_le_bytes(buf);
+
+        Ok(Self { kind, len })
+    }
+}
+
+pub(crate) fn decode(
+    data: &mut impl std::io::Read,
+    compression: Compression,
+) -> Result<(u64, Option<LogMsg>), DecodeError> {
+    use re_protos::external::prost::Message;
+    use re_protos::log_msg::v0::{ArrowMsg, BlueprintActivationCommand, Encoding, SetStoreInfo};
+
+    let mut read_bytes = 0u64;
+    let header = MessageHeader::decode(data)?;
+    read_bytes += std::mem::size_of::<MessageHeader>() as u64 + header.len as u64;
+
+    let mut buf = vec![0; header.len as usize];
+    data.read_exact(&mut buf[..])?;
+
+    let msg = match header.kind {
+        MessageKind::SetStoreInfo => {
+            let set_store_info = SetStoreInfo::decode(&buf[..])?;
+            Some(LogMsg::SetStoreInfo(set_store_info.try_into()?))
+        }
+        MessageKind::ArrowMsg => {
+            let arrow_msg = ArrowMsg::decode(&buf[..])?;
+            if arrow_msg.encoding() != Encoding::ArrowIpc {
+                return Err(DecodeError::Codec(CodecError::UnsupportedEncoding));
+            }
+
+            let (schema, chunk) = decode_arrow(&arrow_msg.payload, compression)?;
+
+            let store_id: re_log_types::StoreId = arrow_msg
+                .store_id
+                .ok_or_else(|| {
+                    TypeConversionError::missing_field("rerun.log_msg.v0.ArrowMsg", "store_id")
+                })?
+                .into();
+
+            let chunk = re_chunk::Chunk::from_transport(&re_chunk::TransportChunk {
+                schema,
+                data: chunk,
+            })?;
+
+            Some(LogMsg::ArrowMsg(store_id, chunk.to_arrow_msg()?))
+        }
+        MessageKind::BlueprintActivationCommand => {
+            let blueprint_activation_command = BlueprintActivationCommand::decode(&buf[..])?;
+            Some(LogMsg::BlueprintActivationCommand(
+                blueprint_activation_command.try_into()?,
+            ))
+        }
+        MessageKind::End => None,
+    };
+
+    Ok((read_bytes, msg))
+}
+
+fn decode_arrow(
+    data: &[u8],
+    compression: crate::Compression,
+) -> Result<
+    (
+        arrow2::datatypes::Schema,
+        arrow2::chunk::Chunk<Box<dyn re_chunk::Arrow2Array>>,
+    ),
+    DecodeError,
+> {
+    let mut uncompressed = Vec::new();
+    let data = match compression {
+        crate::Compression::Off => data,
+        crate::Compression::LZ4 => {
+            lz4_flex::block::decompress_into(data, &mut uncompressed)?;
+            uncompressed.as_slice()
+        }
+    };
+
+    Ok(read_arrow_from_bytes(&mut &data[..])?)
+}
+
+/// Helper function that deserializes raw bytes into arrow schema and record batch
+/// using Arrow IPC format.
+fn read_arrow_from_bytes<R: std::io::Read>(
```

Comment:
re: `decoder` `encoder` being in two places, that's because we still have to continue supporting msgpack for the time being. So all the new protobuf encoding stuff is under `protobuf`.

re: `codec`, it's there separately because that's the format used by gRPC, with its own `MessageHeader`. the existing `encoder` and `decoder` will eventually only be used for files (either local or served over HTTP), so I thought it'd be better to keep them separate.

The naming could be clearer, I can reshuffle the modules to make it more obvious what's going on.
When it comes to the duplication of arrow serialization, it could be in a common module, now that we've decided (have we?) to send the schema separately with each message.

Reply:
yeah, I think just some naming reshuffling might help to make things more obvious. Perhaps even worth splitting grpc ``codec`` into encoder and decoder just for consistency...